numpy/core/src/umath/special_comparisons.cpp

```diff
@@ -0,0 +1,464 @@
+        npy_intp const strides[], NpyAuxData *NPY_UNUSED(auxdata))
+{
+    npy_intp N = dimensions[0];
+    char *out = data[2];
+    npy_intp stride = strides[2];
+
+    while (N--) {
+        *reinterpret_cast<npy_bool *>(out) = result;
+        out += stride;
+    }
+    return 0;
+}
+
+static inline void
+get_min_max(int typenum, long long *min, unsigned long long *max)
+{
+    *min = 0;
+    switch (typenum) {
+        case NPY_BYTE:
+            *min = NPY_MIN_BYTE;
+            *max = NPY_MAX_BYTE;
+            break;
+        case NPY_UBYTE:
+            *max = NPY_MAX_UBYTE;
+            break;
+        case NPY_SHORT:
+            *min = NPY_MIN_SHORT;
+            *max = NPY_MAX_SHORT;
+            break;
+        case NPY_USHORT:
+            *max = NPY_MAX_USHORT;
+            break;
+        case NPY_INT:
+            *min = NPY_MIN_INT;
+            *max = NPY_MAX_INT;
+            break;
+        case NPY_UINT:
+            *max = NPY_MAX_UINT;
+            break;
+        case NPY_LONG:
+            *min = NPY_MIN_LONG;
+            *max = NPY_MAX_LONG;
+            break;
+        case NPY_ULONG:
+            *max = NPY_MAX_ULONG;
+            break;
+        case NPY_LONGLONG:
+            *min = NPY_MIN_LONGLONG;
+            *max = NPY_MAX_LONGLONG;
+            break;
+        case NPY_ULONGLONG:
+            *max = NPY_MAX_ULONGLONG;
+            break;
+        default:
+            assert(0);
+    }
+}
+
+
+/*
+ * Determine if a Python long is within the typenums range, smaller, or larger.
+ * 
+ * Function returns -2 for errors.
+ */
+static int
+get_value_range(PyObject *value, int type_num)
+{
+    long long min;
+    unsigned long long max;
+    get_min_max(type_num, &min, &max);
+
+    int overflow;
+    long long val = PyLong_AsLongLongAndOverflow(value, &overflow);
+    if (val == -1 && overflow == 0 && PyErr_Occurred()) {
+        return (NPY_CASTING)-1;
+    }
+
+    if (overflow == 0) {
+        if (val < min) {
+            return -1;
+        }
+        else if (val > 0 && (unsigned long long)val > max) {
+            return 1;
+        }
+        return 0;
+    }
+    else if (overflow < 0) {
+        return -1; 
+    }
+    else if (max <= NPY_MAX_LONGLONG) {
+        return 1;
+    }
+
+    /*
+     * If we are checking for unisgned long long, the value may be larger
+     * then long long, but within range of unsigned long long.  Check this
+     * by doing the normal Python integer comparison.
+     */
+    PyObject *obj = PyLong_FromUnsignedLongLong(max);
+    if (obj == NULL) {
+        return -2;
+    }
+    int cmp = PyObject_RichCompareBool(value, obj, Py_GT);
+    Py_DECREF(obj);
+    if (cmp < 0) {
+        return -2;
+    }
+    if (cmp) {
+        return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+
+/*
+ * Find the type resolution for any numpy_int with pyint comparison.  This
+ * function supports *both* directions for all types.
+ */
+static NPY_CASTING
+resolve_descriptors_raw(
```

Comment:
Thanks, never liked it either! just didn't have a good idea.  Maybe just `_scalars` or even `_with_scalars`?  Because I think it may be nice to allow that in some future.